pkg/vfs/fsck.go

```diff
@@ -0,0 +1,133 @@
+	TypeMismatch
+)
+
+// FsckLog is a struct for an inconsistency in the VFS
+type FsckLog struct {
+	Type        FsckLogType
+	FileDoc     *FileDoc
+	DirDoc      *DirDoc
+	Filename    string
+	PruneAction string
+	PruneError  error
+}
+
+// String returns a string describing the FsckLog
+func (f *FsckLog) String() string {
+	switch f.Type {
+	case FileMissing:
+		if f.IsFile() {
+			return "the file is present in the index but not on the filesystem"
+		}
+		return "the directory is present in the index but not on the filesystem"
+	case TypeMismatch:
+		if f.IsFile() {
+			return "it's a file in the index but a directory on the filesystem"
+		}
+		return "it's a directory in the index but a file on the filesystem"
+	case IndexMissing:
+		return "the document is present on the local filesystem but not in the index"
+	}
+	panic("bad FsckLog type")
+}
+
+// MarshalJSON implements the json.Marshaler interface
+func (f *FsckLog) MarshalJSON() ([]byte, error) {
+	v := map[string]string{
+		"filename": f.Filename,
+		"message":  f.String(),
+	}
+	if f.IsFile() {
+		v["file_id"] = f.FileDoc.ID()
+	} else {
+		v["file_id"] = f.DirDoc.ID()
+	}
+	if f.PruneAction != "" {
+		v["prune_action"] = f.PruneAction
+		v["prune_error"] = f.PruneError.Error()
+	}
+	return json.Marshal(v)
+}
+
+// IsFile returns whether or not the associated log is for a file document or a
+// directory.
+func (f *FsckLog) IsFile() bool {
+	return f.FileDoc != nil
+}
+
+// FsckPrune tries to fix the given list on inconsistencies in the VFS
+func FsckPrune(fs VFS, logbook []*FsckLog, dryrun bool) {
+	indexer := fs.Index()
+	for _, entry := range logbook {
+		switch entry.Type {
+		case FileMissing:
+			entry.PruneAction = "deleting entry from index"
+			if dryrun {
+				continue
+			}
+			if entry.IsFile() {
+				if err := indexer.DeleteFileDoc(entry.FileDoc); err != nil {
+					entry.PruneError = err
+				}
+			} else {
+				if err := indexer.DeleteDirDoc(entry.DirDoc); err != nil {
+					entry.PruneError = err
+				}
+			}
+		case IndexMissing:
+			if !entry.IsFile() {
+				continue
+			}
+			fileDoc := entry.FileDoc
+			var orphan bool
+			parentDir, err := indexer.DirByID(fileDoc.DirID)
+			if err != nil {
+				orphan = true
+			} else {
+				fullpath := path.Join(parentDir.Fullpath, fileDoc.Name())
+				if _, err := indexer.FileByPath(fullpath); err == nil {
+					entry.PruneError = err
+					continue
+				}
+			}
+			if orphan {
+				entry.PruneAction = "creating index entry in orphan directory"
+			} else {
+				entry.PruneAction = "creating index entry in-place"
+			}
+			if dryrun {
+				continue
+			}
+			if orphan {
+				orphanDir, err := MkdirAll(fs, OrphansDirName, nil)
+				if err != nil {
+					entry.PruneError = err
+					continue
+				}
+				fileDoc.DirID = orphanDir.ID()
+			}
+			if err := indexer.CreateFileDoc(fileDoc); err != nil {
+				entry.PruneError = err
+			}
+		case TypeMismatch:
+			// TODO: determine how to handle type mismatch... This probably require a
```

Comment:
What are the cases that lead here?

- If we have a directory in couchdb and a file in swift, we should keep the directory and recreate the file under a new name (like for a conflict).
- If we have a file in couchdb and a directory, we should delete the file entry and create a directory entry in couch.
- Others?